File: 416.分割等和子集.py
#!/usr/bin/env python
# coding=utf-8
'''
Description: 
Author: yangyuxiang
Date: 2021-04-30 16:02:25
LastEditors: yangyuxiang
LastEditTime: 2021-04-30 17:59:17
FilePath: /leetcode/416.分割等和子集.py
'''
#
# @lc app=leetcode.cn id=416 lang=python
#
# [416] 分割等和子集
#

# @lc code=start
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        n = len(nums)
        if n <= 1:
            return False
        
        s = sum(nums)
        if s % 2 == 1: # 奇数肯定不可分割成和相等的两部分
            return False

        s = s // 2
        # dp = [[False] * (s+1) for _ in range(n+1)]
        # for i in range(len(dp)):
        #     dp[i][0] = True
        
        # for i in range(1, n+1):
        #     for j in range(1, s+1):
        #         if j - nums[i-1] < 0:
        #             dp[i][j] = dp[i-1][j]
        #         else:
        #             dp[i][j] = dp[i-1][j] or dp[i-1][j-nums[i-1]]

        dp = [False] * (s+1)
        dp[0] = True

        for i in range(0, n):
            for j in range(s, -1, -1):
                if j - nums[i] >= 0:
                    dp[j] = dp[j] or dp[j-nums[i]]


        return dp[s]

        
        
        
        
        
# @lc code=end

File: 416.分割等和子集.java
/*
 * @lc app=leetcode.cn id=416 lang=java
 *
 * [416] 分割等和子集
 */

// @lc code=start
class CanPartition {
    public static void main(String[] args) {
        int[] nums = {1,5,10,6};
        System.out.println(new CanPartition().canPartition1(nums));
    }

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
        }
        if (sum % 2 == 1) {
            return false;
        }
        int target = sum / 2;

        // 定义二维dp数组
        int[][] dp = new int[nums.length][target+1];
        // 初始化dp数组
        for (int i = 0; i < nums.length; i++) {
            dp[i][0] = 1;
        }
        for (int j = 1; j < target+1; j++) {
            if (j == nums[0]) {
                dp[0][j] = 1;
            } else {
                dp[0][j] = 0;
            }
        }

        for (int i = 1; i < nums.length; i++) {
            for (int j = 1; j < target+1; j++) {
                if (j - nums[i] >= 0) {
                    if (dp[i-1][j] == 1 || dp[i-1][j-nums[i]] == 1) {
                        dp[i][j] = 1;
                    }
                } else {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }

        return dp[nums.length-1][target] == 1 ? true : false;
    }
    public boolean canPartition1(int[] nums) {
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
        }
        if (sum % 2 == 1) {
            return false;
        }
        int target = sum / 2;
        // 定义一维dp数组
        int[] dp = new int[target+1];
        dp[0] = 1;
        for (int i = 0; i < nums.length; i++) {
            for (int j = target; j > 0; j--) {
                if (j - nums[i] >= 0) {
                    if (dp[j] == 1 || dp[j - nums[i]] == 1) {
                        dp[j] = 1;
                    }
                }
            }
        }

        return dp[target] == 1 ? true : false;
    }
}
// @lc code=end

